In this notebook we will test the candidate generation module 

In [7]:
from src.llm.llm_interface import UnifiedLLMInterface
# IMPORTs
from utils.task import Task
import json
import os
from src.pipeline.candidate_generation import candidate_generation
from dotenv import load_dotenv
from utils.prompt import load_prompt
import tiktoken
from tqdm import tqdm

load_dotenv()

True

In [8]:
# Function to load JSON data
def load_json_data(filepath):
    with open(filepath, 'r') as file:
        data = json.load(file)
    return data


# Function to create task object
def create_task(example):
    return Task(example)

In [9]:
# load the task data
filepath = "C:\\Users\yousf\Bureau\ConvergenceAI\CHESS_Impl\data\\subsampled_dev_set\\sub_sampled_bird_dev_set.json"
data = load_json_data(filepath)
# load the retrieved entities
filepath_entities = "C:/Users\yousf\Bureau\ConvergenceAI\CHESS_Impl\data/test/retrieved_entities.json"
retrieved_entities = load_json_data(filepath_entities)
# load the retrieved context
filepath_context = "C:/Users\yousf\Bureau\ConvergenceAI\CHESS_Impl\data/test/retrieved_context.json"
retrieved_context = load_json_data(filepath_context)
# load the selected schema
filepath_schema = "C:/Users\yousf\Bureau\ConvergenceAI\CHESS_Impl\data/test/selected_schema.json"
selected_schema = load_json_data(filepath_schema)

llm = UnifiedLLMInterface()

In [14]:
# test candidate generation module with a random sample
index = 22
example = data[index]
task = create_task(example)
entity_retrieval = {
    "similar_columns": {
        "bond": [
            "bond_type"
        ]
    },
    "similar_values": {}
}
context_retrieval = {
    "schema_with_descriptions": {
        "bond": {
            "bond_type": {
                "column_name": "",
                "column_description": "type of the bond",
                "value_description": "\u0093-\u0094: single bond '=': double bond '#': triple bond"
            },
            "molecule_id": {
                "column_name": "",
                "column_description": "identifying the molecule in which the bond appears",
                "value_description": ""
            },
            "bond_id": {
                "column_name": "",
                "column_description": "unique id representing bonds",
                "value_description": "TRxxx_A1_A2: TRXXX refers to which molecule A1 and A2 refers to which atom"
            }
        },
        "connected": {
            "bond_id": {
                "column_name": "bond id",
                "column_description": "bond id representing bond between two atoms",
                "value_description": ""
            }
        },
        "atom": {
            "molecule_id": {
                "column_name": "molecule id",
                "column_description": "identifying the molecule to which the atom belongs",
                "value_description": "TRXXX_i represents ith atom of molecule TRXXX"
            }
        }
    }
}
schema = {
    "bond": [
        "bond_type",
        "molecule_id"
    ],
    "molecule": [
        "molecule_id"
    ]
}

model = "llama-3"
ans = candidate_generation(task=task, retrieved_entities=entity_retrieval,
                           retrieved_context=context_retrieval,
                           selected_schema=schema,
                           llm=llm,
                           model=model,
                           num_samples=1)
ans

{
    "chain_of_thought_reasoning": "First, I identified the relevant table as 'bond' since it contains information about bond types. Then, I focused on the 'bond_type' column, which has an example value '#', indicating a triple bond. Next, I realized that I need to count the number of compounds that form a triple bond and then calculate the percentage. To do this, I joined the 'bond' table with the 'olecule' table to get the total number of compounds. Finally, I used the COUNT and SUM functions to calculate the percentage.",
    "SQL": "SELECT ROUND(COUNT(T1.bond_id) / (SELECT COUNT(*) FROM T1) * 100, 2) AS percentage FROM bond T1 WHERE T1.bond_type = '#';"
}


{'chain_of_thought_reasoning': "First, I identified the relevant table as 'bond' since it contains information about bond types. Then, I focused on the 'bond_type' column, which has an example value '#', indicating a triple bond. Next, I realized that I need to count the number of compounds that form a triple bond and then calculate the percentage. To do this, I joined the 'bond' table with the 'olecule' table to get the total number of compounds. Finally, I used the COUNT and SUM functions to calculate the percentage.",
 'SQL': "SELECT ROUND(COUNT(T1.bond_id) / (SELECT COUNT(*) FROM T1) * 100, 2) AS percentage FROM bond T1 WHERE T1.bond_type = '#';"}

In [5]:
#test candidate generation module with the subsampled test
model = "gpt-4"
num_samples = 1
res = []
for index, sample in tqdm(enumerate(data)):
    task = create_task(sample)
    ans = candidate_generation(task=task, retrieved_entities=retrieved_entities[index],
                               retrieved_context=retrieved_context[index],
                               selected_schema=selected_schema[index]["selected_schema"],
                               model=model,
                               num_samples=num_samples)
    res.append(ans)
    print(str(index) + " : ", ans)

1it [00:11, 11.06s/it]

0 :  {'chain_of_thought_reasoning': "To answer the question, I need to perform two tasks: 1) Identify the atom IDs that belong to the molecule 'TR346'. This can be achieved by querying the 'atom' table where 'molecule_id' equals 'TR346'. 2) Count the distinct types of bonds that can be created by the molecule 'TR346'. This requires querying the 'bond' table where 'molecule_id' equals 'TR346' and counting the distinct 'bond_type'. Both results need to be obtained in a single query. To achieve this, I will use a JOIN operation between the 'atom' and 'bond' tables on the 'molecule_id' column, filtering both tables for 'molecule_id' = 'TR346'. The final query will select the 'atom_id' from the 'atom' table and count the distinct 'bond_type' from the 'bond' table.", 'SQL': "SELECT T1.atom_id, COUNT(DISTINCT T2.bond_type) FROM atom T1 JOIN bond T2 ON T1.molecule_id = T2.molecule_id WHERE T1.molecule_id = 'TR346' AND T2.molecule_id = 'TR346' GROUP BY T1.atom_id"}


2it [00:21, 10.84s/it]

1 :  {'chain_of_thought_reasoning': "To answer the question, I need to find the city of a high school in the county of Merced with specific criteria. First, I'll join the 'schools' and 'frpm' tables on the 'CDSCode' since it's a common key and a foreign key relationship exists between these tables. I'll filter the 'schools' table for schools in Merced County and with an EILCode of 'HS' for high schools. Then, I'll filter the 'frpm' table for records with 'NSLP Provision Status' as 'Lunch Provision 2' and where the lowest grade is 9 and the highest grade is 12. The final step is to select the 'City' from the 'schools' table which meets all these conditions.", 'SQL': "SELECT T1.City FROM schools T1 JOIN frpm T2 ON T1.CDSCode = T2.CDSCode WHERE T1.County = 'Merced' AND T1.EILCode = 'HS' AND T2.`NSLP Provision Status` = 'Lunch Provision 2' AND T2.`Low Grade` = '9' AND T2.`High Grade` = '12'"}


3it [00:31, 10.18s/it]

2 :  {'chain_of_thought_reasoning': "To find the full name of the tallest hero published by Marvel Comics, I need to join the 'superhero' table with the 'publisher' table based on the foreign key relationship between superhero.publisher_id and publisher.id. I will filter superheroes by those published by 'Marvel Comics', which is specified in the publisher_name column of the publisher table. To ensure I get the tallest hero, I will order the results by height in descending order and select the top result using LIMIT 1. I will also ensure to filter out any superheroes with a height of NULL or 0 to adhere to the database admin instructions. The final query will only select the full_name of the superhero, as per the instructions.", 'SQL': "SELECT T1.full_name FROM superhero T1 JOIN publisher T2 ON T1.publisher_id = T2.id WHERE T2.publisher_name = 'Marvel Comics' AND T1.height_cm IS NOT NULL AND T1.height_cm > 0 ORDER BY T1.height_cm DESC LIMIT 1"}


4it [00:44, 11.27s/it]

3 :  {'chain_of_thought_reasoning': "To answer the question, I need to count the number of users who are classified as elders (Age > 65) and have obtained the 'Supporter' badge. I will use the 'users' table to filter users who are elders and the 'badges' table to filter badges with the name 'Supporter'. I will join these tables on the UserId to ensure that the badges are associated with the correct users. The COUNT function will be used to count the number of such users, and I will ensure to filter out any null values in the UserId and Age columns to adhere to the database admin instructions. The final query will join the tables, apply the necessary filters, and count the distinct users meeting the criteria.", 'SQL': "SELECT COUNT(DISTINCT T1.UserId) FROM badges T1 JOIN users T2 ON T1.UserId = T2.Id WHERE T2.Age > 65 AND T1.Name = 'Supporter' AND T1.UserId IS NOT NULL AND T2.Age IS NOT NULL"}


5it [00:55, 11.31s/it]

4 :  {'chain_of_thought_reasoning': "To solve the query, I need to count users from 'New York' who have both 'Teacher' and 'Supporter' badges. First, I'll join the 'users' table with the 'badges' table on the user ID. Since the question specifies users from 'New York', I'll filter the 'Location' to 'New York'. To ensure a user has both 'Teacher' and 'Supporter' badges, I'll use a GROUP BY on the user ID and HAVING clause to check the presence of both badges. The COUNT function will then give the number of such users. I'll use aliases for tables as per the instructions and ensure to filter out any null values in relevant columns.", 'SQL': "SELECT COUNT(DISTINCT T1.Id) FROM users T1 JOIN badges T2 ON T1.Id = T2.UserId WHERE T1.Location = 'New York' AND T2.Name IS NOT NULL GROUP BY T1.Id HAVING SUM(CASE WHEN T2.Name = 'Teacher' THEN 1 ELSE 0 END) > 0 AND SUM(CASE WHEN T2.Name = 'Supporter' THEN 1 ELSE 0 END) > 0"}


6it [01:06, 11.13s/it]

5 :  {'chain_of_thought_reasoning': "To answer the question about the description of the ruling for the card named 'Condemn', I need to join the 'cards' table with the 'rulings' table based on the 'uuid' which is a common key between the two tables. The 'cards' table contains the card names and their corresponding UUIDs, while the 'rulings' table contains the text of the rulings linked to these UUIDs. The query will filter the 'cards' table to find the UUID of the card named 'Condemn', then use this UUID to fetch the corresponding ruling text from the 'rulings' table. The SELECT clause will only include the 'text' column from the 'rulings' table as per the instructions to output only the information asked in the question.", 'SQL': "SELECT T2.text FROM cards T1 JOIN rulings T2 ON T1.uuid = T2.uuid WHERE T1.name = 'Condemn'"}


7it [01:16, 10.74s/it]

6 :  {'chain_of_thought_reasoning': "To answer the question, I need to join the 'molecule' and 'bond' tables based on the 'molecule_id' to access both the carcinogenic status of the molecule and the type of bonds it contains. The molecule of interest is identified by 'molecule_id' = 'TR006'. I will count the number of bonds where 'bond_type' = '=' (double bonds) for this molecule. The SQL query will use a JOIN operation between the two tables, filter by 'molecule_id', and count the number of double bonds. The carcinogenic status will be directly selected from the 'label' column of the 'molecule' table. The query will ensure that only the required information (count of double bonds and carcinogenic status) is returned, following the database admin instructions.", 'SQL': "SELECT COUNT(T2.bond_id) AS double_bonds_count, T1.label AS carcinogenic FROM molecule T1 JOIN bond T2 ON T1.molecule_id = T2.molecule_id WHERE T1.molecule_id = 'TR006' AND T2.bond_type = '='"}


8it [01:26, 10.66s/it]

7 :  {'chain_of_thought_reasoning': "To answer the question, I need to join the 'member' table with the 'major' table using the foreign key relationship between 'member.link_to_major' and 'major.major_id'. The question specifically asks for the full name of the member and the college they belong to, where the member's position is 'Secretary'. According to the schema, the full name is a combination of 'first_name' and 'last_name' from the 'member' table, and the college is in the 'major' table. I will use an INNER JOIN to ensure that only members with a valid major are included. The query will filter members by the position 'Secretary', and since the question does not specify sorting or the need to handle multiple secretaries, I will assume there is only one or the query should return all matching entries. I will use aliases for tables as instructed and ensure to select only the required columns.", 'SQL': "SELECT T1.first_name || ' ' || T1.last_name AS full_name, T2.college FROM member 

9it [01:39, 11.26s/it]

8 :  {'chain_of_thought_reasoning': "To determine the percentage of compounds that form a triple bond, we need to calculate two main values: the total number of unique molecules and the number of unique molecules that have at least one triple bond. We can find the total number of unique molecules by counting distinct molecule_id in the molecule table. For the number of molecules with at least one triple bond, we can count distinct molecule_id in the bond table where bond_type is '#'. After obtaining these two counts, the percentage can be calculated by dividing the number of molecules with a triple bond by the total number of molecules and then multiplying by 100. We will use a LEFT JOIN to ensure all molecules are considered, even those without any bonds. The use of COALESCE will handle cases where there are no triple bonds, ensuring the calculation defaults to zero in such cases.", 'SQL': "SELECT (COUNT(DISTINCT T2.molecule_id) * 100.0 / COUNT(DISTINCT T1.molecule_id)) AS percentage_

10it [01:48, 10.67s/it]

9 :  {'chain_of_thought_reasoning': "To answer the question about Francesco Parravicini's potential on a specific date, I need to join the Player and Player_Attributes tables. The Player table will help identify the player_api_id for Francesco Parravicini, and the Player_Attributes table contains the potential scores. I will use the player_name from the Player table to filter for 'Francesco Parravicini' and the date from the Player_Attributes table to filter for '2010-08-30 00:00:00'. I will join these tables on player_api_id since it's a common key in both tables. The query will select the potential column from the Player_Attributes table, ensuring to filter out any potential null values in the potential column as per the database admin instructions.", 'SQL': "SELECT T2.potential FROM Player T1 JOIN Player_Attributes T2 ON T1.player_api_id = T2.player_api_id WHERE T1.player_name = 'Francesco Parravicini' AND T2.date = '2010-08-30 00:00:00' AND T2.potential IS NOT NULL"}


11it [01:57, 10.23s/it]

10 :  {'chain_of_thought_reasoning': "To answer the question, I need to count the number of male clients from the 'client' table who live in the 'North Bohemia' region and have an average salary greater than 8000. This requires joining the 'client' table with the 'district' table on the 'district_id' to access the region and average salary information. I will filter the clients based on gender ('M' for male) and use the conditions specified for the region ('North Bohemia') and the average salary (>8000). The SQL query will use a JOIN operation to combine the tables, a WHERE clause to filter the data, and a COUNT function to get the number of clients meeting the criteria.", 'SQL': "SELECT COUNT(*) FROM client T1 JOIN district T2 ON T1.district_id = T2.district_id WHERE T1.gender = 'M' AND T2.A3 = 'North Bohemia' AND T2.A11 > 8000"}


12it [02:08, 10.31s/it]

11 :  {'chain_of_thought_reasoning': "To answer the question, I need to find the school in California ('CA') with the lowest latitude. The database schema indicates that the 'schools' table contains all necessary information. I will filter schools based on the 'State' column to include only those in California. Then, I will sort these schools by the 'Latitude' column in ascending order to find the school with the lowest latitude. Since the latitude can be null, I will filter out schools where latitude is null. The question asks for the city and the lowest grade of this school, as well as the school name. I will use the 'City', 'GSserved', and 'School' columns in my SELECT clause. I will limit the results to 1 to get only the school with the absolute lowest latitude. I will not use concatenation in the SELECT clause, as per the instructions.", 'SQL': "SELECT T1.City, T1.GSserved, T1.School FROM schools T1 WHERE T1.State = 'CA' AND T1.Latitude IS NOT NULL ORDER BY T1.Latitude ASC LIMIT 1

13it [02:19, 10.43s/it]

12 :  {'chain_of_thought_reasoning': "To solve the problem, I need to calculate the percentage of bonds that are double bonds ('=') among all bonds in molecule TR047. First, I'll filter the bonds by molecule_id = 'TR047'. Then, I'll count the total number of bonds and the number of double bonds for this molecule. The percentage is calculated by dividing the number of double bonds by the total number of bonds and multiplying by 100 to get a percentage. I'll use a single SQL query with a CASE statement to conditionally count double bonds and use COUNT to count all bonds. The result will be grouped by molecule_id, though this is implicit since we're only considering TR047.", 'SQL': "SELECT (SUM(CASE WHEN bond_type = '=' THEN 1 ELSE 0 END) * 100.0 / COUNT(bond_id)) AS percent FROM bond WHERE molecule_id = 'TR047'"}


14it [02:28, 10.22s/it]

13 :  {'chain_of_thought_reasoning': "To answer the question, I need to find the user who added a bounty of 50 to a post with a title mentioning 'variance'. First, I'll identify the relevant tables and columns: 'votes' for BountyAmount, 'posts' for Title, and 'users' for DisplayName. I'll join these tables based on the relationships defined in the schema: votes.PostId to posts.Id and votes.UserId to users.Id. I'll filter the votes table for BountyAmount = 50 and the posts table for titles containing 'variance'. I'll use a LIKE operator with '%' wildcards to handle any position of 'variance' in the title. Finally, I'll select the DisplayName of the user, ensuring to filter out any null values in the BountyAmount and Title columns as per the admin instructions.", 'SQL': "SELECT T3.DisplayName FROM votes T1 JOIN posts T2 ON T1.PostId = T2.Id JOIN users T3 ON T1.UserId = T3.Id WHERE T1.BountyAmount = 50 AND T2.Title LIKE '%variance%' AND T1.BountyAmount IS NOT NULL AND T2.Title IS NOT NULL

15it [02:38, 10.55s/it]

14 :  {'chain_of_thought_reasoning': "To answer the question, I need to join the 'users' and 'badges' tables based on the user's ID. The 'users' table contains the location information, and the 'badges' table contains the badge acquisition dates. I will filter the 'users' table to include only those users located in 'Rochester, NY'. Then, I will join this filtered list with the 'badges' table to get the dates when these users obtained their badges. I will use aliases for the tables to make the query clearer and ensure that I only select the 'Date' column from the 'badges' table as per the instructions. I will also ensure to filter out any null values in the 'Date' column to comply with the admin instructions.", 'SQL': "SELECT T2.Date FROM users T1 JOIN badges T2 ON T1.Id = T2.UserId WHERE T1.Location = 'Rochester, NY' AND T2.Date IS NOT NULL"}


In [6]:
from pathlib import Path

schema_path = Path('C:/Users/yousf/Bureau/ConvergenceAI/CHESS_Impl/data/test/generated_candidate.json')
with open(schema_path, 'w') as f:
    json.dump(res, f, indent=4)

## Cost Estimation per task

In [5]:

PROMPT_PATH = os.getenv("PROMPT_ROOT_PATH") + "\\candidate_generation.txt"
prompt = load_prompt(PROMPT_PATH)

In [6]:

def tokens_calc(example):
    encoding = tiktoken.get_encoding("cl100k_base")
    num_tokens = len(encoding.encode(example))
    return num_tokens

In [7]:
#prompt template tokens 
tokens_calc(prompt)

631

The prompt template has 631 tokens in total, and we have also 3 other variables (Database_Schema,Question and Hint).
After i see a formatted prompt example it contains about 2000 tokens (because database_schema will be long)
So let's suppose that input tokens is 2000

In [8]:
## Output tokens estimation from an example 
output_example = """
{'chain_of_thought_reasoning': "First, I identified the relevant columns from the question: CDSCode, County Code, School Type, Low Grade, High Grade, and County Name. Then, I determined that the question is asking for the city location of a high school level school with Lunch Provision 2, whose lowest grade is 9 and the highest grade is 12 in the county of Merced. I used the hint that high school can be represented as EILCode = 'HS'. I joined the frpm and schools tables based on the CDSCode, and then filtered the results to match the conditions specified in the question. Finally, I selected the City column from the schools table, which is the column that provides the city location.",
 'SQL': "SELECT City FROM schools WHERE EILCode = 'HS' AND County = 'Merced' AND LowGrade = '9' AND HighGrade = '12' AND NSLPProvisionStatus = 'Lunch Provision 2' AND CDSCode IN (SELECT CDSCode FROM frpm WHERE CountyCode = '02');"}
"""
tokens_calc(output_example)

227

Let's suppose the output tokens is 250

In [9]:
## Price calculation (just with gpt4 because in this module we don't use gpt3.5) 
input_price_per_token_gpt4 = 0.01 / 1000
output_price_per_token_gpt4 = 0.03 / 1000
price_gpt4 = 2000 * input_price_per_token_gpt4 + 250 * output_price_per_token_gpt4
print("estimated price per Task (GPT-4):", price_gpt4, "$")

estimated price per Task (GPT-4): 0.0275 $


In [10]:
## in this module there is a number of retrials so let's estimate the price with different number of retrials 

num_retrials = [1, 2, 3, 4, 5]
for num_retrial in num_retrials:
    total_price = price_gpt4 * num_retrial
    print("The estimated price with number of retrials of " + str(num_retrial) + ": ", total_price)

The estimated price with number of retrials of 1:  0.0275
The estimated price with number of retrials of 2:  0.055
The estimated price with number of retrials of 3:  0.0825
The estimated price with number of retrials of 4:  0.11
The estimated price with number of retrials of 5:  0.1375
